File: FBControlCoreTests/Tests/Integration/FBFileReaderTests.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <XCTest/XCTest.h>

#import <FBControlCore/FBControlCore.h>

#import <sys/types.h>
#import <sys/stat.h>

@interface FBFileReaderTests : XCTestCase <FBDataConsumer>

@property (atomic, assign, readwrite) BOOL didRecieveEOF;

@end

@implementation FBFileReaderTests

- (void)setUp
{
  self.didRecieveEOF = NO;
}

- (void)testConsumesData
{
  // Setup
  NSPipe *pipe = NSPipe.pipe;
  id<FBAccumulatingBuffer> consumer = FBDataBuffer.accumulatingBuffer;
  FBFileReader *reader = [FBFileReader readerWithFileDescriptor:pipe.fileHandleForReading.fileDescriptor closeOnEndOfFile:NO consumer:consumer logger:nil];
  XCTAssertEqual(reader.state, FBFileReaderStateNotStarted);

  // Start reading
  NSError *error = nil;
  BOOL success = [[reader startReading] await:&error] != nil;
  XCTAssertNil(error);
  XCTAssertTrue(success);
  XCTAssertEqual(reader.state, FBFileReaderStateReading);

  // Write some data and confirm that it is as expected.
  NSData *expected = [@"Foo Bar Baz" dataUsingEncoding:NSUTF8StringEncoding];
  [pipe.fileHandleForWriting writeData:expected];
  [pipe.fileHandleForWriting closeFile];
  NSPredicate *predicate = [NSPredicate predicateWithBlock:^ BOOL (id _, id __) {
    return [expected isEqualToData:consumer.data];
  }];
  XCTestExpectation *expectation = [self expectationForPredicate:predicate evaluatedWithObject:self handler:nil];
  [self waitForExpectations:@[expectation] timeout:FBControlCoreGlobalConfiguration.fastTimeout];

  // Stop reading
  NSNumber *result = [[reader stopReading] await:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(result, @0);
  XCTAssertEqualObjects(reader.finishedReading.result, @0);
  XCTAssertEqual(reader.state, FBFileReaderStateFinishedReadingNormally);
}

- (void)testConsumesEOFAfterStoppedReading
{
  // Setup
  NSPipe *pipe = NSPipe.pipe;
  FBFileReader *reader = [FBFileReader readerWithFileDescriptor:pipe.fileHandleForReading.fileDescriptor closeOnEndOfFile:NO consumer:self logger:nil];
  XCTAssertEqual(reader.state, FBFileReaderStateNotStarted);

  // Start reading
  NSError *error = nil;
  BOOL success = [[reader startReading] await:&error] != nil;
  XCTAssertEqual(reader.state, FBFileReaderStateReading);
  XCTAssertNil(error);
  XCTAssertTrue(success);

  // Write some data.
  NSData *expected = [@"Foo Bar Baz" dataUsingEncoding:NSUTF8StringEncoding];
  [pipe.fileHandleForWriting writeData:expected];

  // Stop reading, we may recieve the consumeEndOfFile on a different queue
  // This is fine as this call will block until the call has happened.
  // Also the assignment is atomic.
  NSNumber *result = [[reader stopReading] await:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(result, @(ECANCELED));
  XCTAssertEqualObjects(reader.finishedReading.result, @(ECANCELED));
  XCTAssertEqual(reader.state, FBFileReaderStateFinishedReadingByCancellation);

  // Confirm we got an eof
  XCTAssertTrue(self.didRecieveEOF);
}

- (void)testConsumesEOFAfterStoppedReadingEvenIfOtherEndOfFifoDoesNotClose
{
  NSString *fifoPath = [NSTemporaryDirectory() stringByAppendingPathComponent:NSUUID.UUID.UUIDString];
  int status = mkfifo(fifoPath.UTF8String, S_IWUSR | S_IRUSR);
  XCTAssertEqual(status, 0);

  NSError *error = nil;
  NSArray<id> *writerAndReader = [[FBFuture
    futureWithFutures:@[
      [FBFileWriter asyncWriterForFilePath:fifoPath],
      [FBFileReader readerWithFilePath:fifoPath consumer:self logger:nil],
    ]]
    await:&error];
  XCTAssertNil(error);
  XCTAssertNotNil(writerAndReader);

  id<FBDataConsumer> writer = writerAndReader[0];
  FBFileReader *reader = writerAndReader[1];

  BOOL success = [[reader startReading] await:&error] != nil;
  XCTAssertNil(error);
  XCTAssertTrue(success);

  // Write some data
  [writer consumeData:[@"HELLO\n" dataUsingEncoding:NSUTF8StringEncoding]];
  [writer consumeData:[@"THERE\n" dataUsingEncoding:NSUTF8StringEncoding]];
  [writer consumeEndOfFile];

  success = [[reader stopReading] await:&error] != nil;
  XCTAssertNil(error);
  XCTAssertTrue(success);

  // Confirm we got an eof
  XCTAssertTrue(self.didRecieveEOF);
}

- (void)testCanStopReadingBeforeEOFResolvesWhenPipeCloses
{
  // Setup
  NSPipe *pipe = NSPipe.pipe;
  id<FBAccumulatingBuffer> consumer = FBDataBuffer.accumulatingBuffer;
  FBFileReader *reader = [FBFileReader readerWithFileDescriptor:pipe.fileHandleForReading.fileDescriptor closeOnEndOfFile:NO consumer:consumer logger:nil];
  XCTAssertEqual(reader.state, FBFileReaderStateNotStarted);

  // Start reading
  NSError *error = nil;
  BOOL success = [[reader startReading] await:&error] != nil;
  XCTAssertNil(error);
  XCTAssertTrue(success);
  XCTAssertEqual(reader.state, FBFileReaderStateReading);

  // Write some data and confirm that it is as expected.
  NSData *expected = [@"Foo Bar Baz" dataUsingEncoding:NSUTF8StringEncoding];
  [pipe.fileHandleForWriting writeData:expected];
  NSPredicate *predicate = [NSPredicate predicateWithBlock:^ BOOL (id _, id __) {
    return [expected isEqualToData:consumer.data];
  }];
  XCTestExpectation *expectation = [self expectationForPredicate:predicate evaluatedWithObject:self handler:nil];
  [self waitForExpectations:@[expectation] timeout:FBControlCoreGlobalConfiguration.fastTimeout];

  // Stop reading, it shouldn't matter that an EOF wasn't sent
  NSNumber *result = [[reader stopReading] await:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(result, @(ECANCELED));
  XCTAssertEqualObjects(reader.finishedReading.result, @(ECANCELED));
  XCTAssertEqual(reader.state, FBFileReaderStateFinishedReadingByCancellation);

  // Write EOF
  [pipe.fileHandleForWriting closeFile];
}

- (void)testPipeClosingBehindBackOfConsumer
{
  // Setup
  NSPipe *pipe = NSPipe.pipe;
  id<FBAccumulatingBuffer> consumer = FBDataBuffer.accumulatingBuffer;
  FBFileReader *reader = [FBFileReader readerWithFileDescriptor:pipe.fileHandleForReading.fileDescriptor closeOnEndOfFile:NO consumer:consumer logger:nil];
  XCTAssertEqual(reader.state, FBFileReaderStateNotStarted);

  // Start reading
  NSError *error = nil;
  BOOL success = [[reader startReading] await:&error] != nil;
  XCTAssertNil(error);
  XCTAssertTrue(success);
  XCTAssertEqual(reader.state, FBFileReaderStateReading);

  // Write some data and confirm that it is as expected.
  NSData *expected = [@"Foo Bar Baz" dataUsingEncoding:NSUTF8StringEncoding];
  [pipe.fileHandleForWriting writeData:expected];
  [pipe.fileHandleForWriting closeFile];
  NSPredicate *predicate = [NSPredicate predicateWithBlock:^ BOOL (id _, id __) {
    return [expected isEqualToData:consumer.data];
  }];
  XCTestExpectation *expectation = [self expectationForPredicate:predicate evaluatedWithObject:self handler:nil];
  [self waitForExpectations:@[expectation] timeout:FBControlCoreGlobalConfiguration.fastTimeout];

  // Stop reading, it shouldn't matter that an EOF wasn't sent
  NSNumber *result = [[reader stopReading] await:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(result, @0);
  XCTAssertEqualObjects(reader.finishedReading.result, @0);
  XCTAssertEqual(reader.state, FBFileReaderStateFinishedReadingNormally);
}

- (void)testReadsFromFilePath
{
  // Read some data.
  NSError *error = nil;
  FBFileReader *reader = [[FBFileReader readerWithFilePath:@"/dev/urandom" consumer:self logger:nil] await:&error];
  XCTAssertNil(error);
  XCTAssertNotNil(reader);
  XCTAssertEqual(reader.state, FBFileReaderStateNotStarted);

  // Start reading
  BOOL success = [[reader startReading] await:&error] != nil;
  XCTAssertNil(error);
  XCTAssertTrue(success);
  XCTAssertEqual(reader.state, FBFileReaderStateReading);

  // Stop Reading
  error = nil;
  NSNumber *result = [[reader stopReading] await:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(result, @(ECANCELED));
  XCTAssertEqualObjects(reader.finishedReading.result, @(ECANCELED));
  XCTAssertEqual(reader.state, FBFileReaderStateFinishedReadingByCancellation);
}

- (void)testReadingTwiceFails
{
  // Read some data.
  NSError *error = nil;
  FBFileReader *reader = [[FBFileReader readerWithFilePath:@"/dev/urandom" consumer:self logger:nil] await:&error];
  XCTAssertNil(error);
  XCTAssertNotNil(reader);
  XCTAssertEqual(reader.state, FBFileReaderStateNotStarted);

  // Start reading.
  BOOL success = [[reader startReading] await:&error] != nil;
  XCTAssertNil(error);
  XCTAssertTrue(success);
  XCTAssertEqual(reader.state, FBFileReaderStateReading);

  // Fail when starting again.
  error = nil;
  success = [[reader startReading] await:&error] != nil;
  XCTAssertNotNil(error);
  XCTAssertEqual(reader.state, FBFileReaderStateReading);

  // Cancellation should work.
  error = nil;
  NSNumber *result = [[reader stopReading] await:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(result, @(ECANCELED));
  XCTAssertEqualObjects(reader.finishedReading.result, @(ECANCELED));
  XCTAssertEqual(reader.state, FBFileReaderStateFinishedReadingByCancellation);
}

- (void)testStoppingTwiceDoesNotError
{
  // Read some data.
  NSError *error = nil;
  FBFileReader *reader = [[FBFileReader readerWithFilePath:@"/dev/urandom" consumer:self logger:nil] await:&error];
  XCTAssertNil(error);
  XCTAssertNotNil(reader);
  XCTAssertEqual(reader.state, FBFileReaderStateNotStarted);

  // Start reading
  BOOL success = [[reader startReading] await:&error] != nil;
  XCTAssertNil(error);
  XCTAssertTrue(success);
  XCTAssertEqual(reader.state, FBFileReaderStateReading);

  // Stop Reading
  error = nil;
  NSNumber *result = [[reader stopReading] await:&error];
  XCTAssertNil(error);
  XCTAssertEqual(reader.state, FBFileReaderStateFinishedReadingByCancellation);
  XCTAssertEqualObjects(result, @(ECANCELED));

  // Stop Reading
  error = nil;
  result = [[reader stopReading] await:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(result, @(ECANCELED));
  XCTAssertEqualObjects(reader.finishedReading.result, @(ECANCELED));
  XCTAssertEqual(reader.state, FBFileReaderStateFinishedReadingByCancellation);
}

- (void)testCancellationOnFinishedReading
{
  // Read some data.
  NSError *error = nil;
  FBFileReader *reader = [[FBFileReader readerWithFilePath:@"/dev/urandom" consumer:self logger:nil] await:&error];
  XCTAssertNil(error);
  XCTAssertNotNil(reader);
  XCTAssertEqual(reader.state, FBFileReaderStateNotStarted);

  // Start reading
  BOOL success = [[reader startReading] await:&error] != nil;
  XCTAssertNil(error);
  XCTAssertTrue(success);
  XCTAssertEqual(reader.state, FBFileReaderStateReading);

  // Stop Reading
  FBFuture<NSNumber *> *finished = reader.finishedReading;
  XCTAssertEqual(finished.state, FBFutureStateRunning);
  success = [[finished cancel] await:&error] != nil;
  XCTAssertNil(error);
  XCTAssertTrue(success);
  XCTAssertEqual(finished.state, FBFutureStateCancelled);
  XCTAssertEqual(reader.state, FBFileReaderStateFinishedReadingByCancellation);
}

- (void)testConcurrentAttachmentIsProhibited
{
  // Read some data.
  NSError *error = nil;
  FBFileReader *reader = [[FBFileReader readerWithFilePath:@"/dev/urandom" consumer:self logger:nil] await:&error];
  XCTAssertNil(error);
  XCTAssertNotNil(reader);
  XCTAssertEqual(reader.state, FBFileReaderStateNotStarted);

  dispatch_queue_t concurrentQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0);
  dispatch_group_t group = dispatch_group_create();
  __block FBFuture<NSNull *> *firstAttempt = nil;
  __block FBFuture<NSNull *> *secondAttempt = nil;
  __block FBFuture<NSNull *> *thirdAttempt = nil;

  dispatch_group_async(group, concurrentQueue, ^{
    firstAttempt = [reader startReading];
  });
  dispatch_group_async(group, concurrentQueue, ^{
    secondAttempt = [reader startReading];
  });
  dispatch_group_async(group, concurrentQueue, ^{
    thirdAttempt = [reader startReading];
  });
  dispatch_group_wait(group, DISPATCH_TIME_FOREVER);

  [firstAttempt await:nil];
  [secondAttempt await:nil];
  [thirdAttempt await:nil];
  XCTAssertEqual(reader.state, FBFileReaderStateReading);

  NSUInteger successes = 0;
  if (firstAttempt.state == FBFutureStateDone) {
    successes++;
  }
  if (secondAttempt.state == FBFutureStateDone) {
    successes++;
  }
  if (thirdAttempt.state == FBFutureStateDone) {
    successes++;
  }

  XCTAssertEqual(successes, 1u);
}

- (void)testAttemptingToReadAGarbageFileDescriptor
{
  // Setup
  FBFileReader *reader = [FBFileReader readerWithFileDescriptor:92123 closeOnEndOfFile:NO consumer:self logger:nil];
  XCTAssertEqual(reader.state, FBFileReaderStateNotStarted);

  // Start reading, the start is asyncrhonous, so we can't know ahead of time if the fd is bad.
  NSError *error = nil;
  BOOL success = [[reader startReading] await:&error] != nil;
  XCTAssertNil(error);
  XCTAssertTrue(success);

  // Write some data and confirm that it is as expected.
  NSNumber *result = [[reader stopReading] await:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(result, @(EBADF));
  XCTAssertEqualObjects(reader.finishedReading.result, @(EBADF));
  XCTAssertEqual(reader.state, FBFileReaderStateFinishedReadingInError);
}

#pragma mark FBDataConsumer Implementation

- (void)consumeEndOfFile
{
  self.didRecieveEOF = YES;
}

- (void)consumeData:(NSData *)data
{
}

@end

File: FBControlCoreTests/Tests/Integration/FBFileReaderTests.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import XCTest

@testable import FBControlCore
import FBControlCoreTestDoubles

final class FBFileReaderTests: XCTestCase, FBDataConsumer {

  var didRecieveEOF: Bool = false

  override func setUp() {
    super.setUp()
    didRecieveEOF = false
  }

  func testConsumesData() throws {
    let pipe = Pipe()
    let consumer = FBDataBuffer.accumulatingBuffer()
    let reader = FBFileReader(fileDescriptor: pipe.fileHandleForReading.fileDescriptor, closeOnEndOfFile: false, consumer: consumer, logger: nil)
    XCTAssertEqual(reader.state, FBFileReaderState.notStarted)

    try reader.startReading().`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.reading)

    let expected = "Foo Bar Baz".data(using: .utf8)!
    pipe.fileHandleForWriting.write(expected)
    pipe.fileHandleForWriting.closeFile()
    let predicate = NSPredicate { _, _ in
      expected == consumer.data()
    }
    let expectation = self.expectation(for: predicate, evaluatedWith: self, handler: nil)
    wait(for: [expectation], timeout: FBControlCoreGlobalConfiguration.fastTimeout)

    let result: NSNumber = try reader.stopReading().`await`()
    XCTAssertEqual(result, 0)
    XCTAssertEqual(reader.finishedReading.result, 0)
    XCTAssertEqual(reader.state, FBFileReaderState.finishedReadingNormally)
  }

  func testConsumesEOFAfterStoppedReading() throws {
    let pipe = Pipe()
    let reader = FBFileReader(fileDescriptor: pipe.fileHandleForReading.fileDescriptor, closeOnEndOfFile: false, consumer: self, logger: nil)
    XCTAssertEqual(reader.state, FBFileReaderState.notStarted)

    try reader.startReading().`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.reading)

    let expected = "Foo Bar Baz".data(using: .utf8)!
    pipe.fileHandleForWriting.write(expected)

    let result: NSNumber = try reader.stopReading().`await`()
    XCTAssertEqual(result, NSNumber(value: ECANCELED))
    XCTAssertEqual(reader.finishedReading.result, NSNumber(value: ECANCELED))
    XCTAssertEqual(reader.state, FBFileReaderState.finishedReadingByCancellation)

    XCTAssertTrue(didRecieveEOF)
  }

  func testConsumesEOFAfterStoppedReadingEvenIfOtherEndOfFifoDoesNotClose() throws {
    let fifoPath = (NSTemporaryDirectory() as NSString).appendingPathComponent(UUID().uuidString)
    let status = mkfifo(fifoPath, S_IWUSR | S_IRUSR)
    XCTAssertEqual(status, 0)

    let writerFuture = FBFileWriter.asyncWriter(forFilePath: fifoPath)
    let readerFuture = FBFileReader.reader(withFilePath: fifoPath, consumer: self, logger: nil)
    let writerAndReader = try FBFutureTestHelpers.combineFutures([writerFuture, readerFuture]).`await`() as NSArray?
    XCTAssertNotNil(writerAndReader)

    // swiftlint:disable force_cast
    let writer = writerAndReader![0] as! FBDataConsumer
    let reader = writerAndReader![1] as! FBFileReader
    // swiftlint:enable force_cast

    try reader.startReading().`await`()

    writer.consumeData("HELLO\n".data(using: .utf8)!)
    writer.consumeData("THERE\n".data(using: .utf8)!)
    writer.consumeEndOfFile()

    try reader.stopReading().`await`()

    XCTAssertTrue(didRecieveEOF)
  }

  func testCanStopReadingBeforeEOFResolvesWhenPipeCloses() throws {
    let pipe = Pipe()
    let consumer = FBDataBuffer.accumulatingBuffer()
    let reader = FBFileReader(fileDescriptor: pipe.fileHandleForReading.fileDescriptor, closeOnEndOfFile: false, consumer: consumer, logger: nil)
    XCTAssertEqual(reader.state, FBFileReaderState.notStarted)

    try reader.startReading().`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.reading)

    let expected = "Foo Bar Baz".data(using: .utf8)!
    pipe.fileHandleForWriting.write(expected)
    let predicate = NSPredicate { _, _ in
      expected == consumer.data()
    }
    let expectation = self.expectation(for: predicate, evaluatedWith: self, handler: nil)
    wait(for: [expectation], timeout: FBControlCoreGlobalConfiguration.fastTimeout)

    let result: NSNumber = try reader.stopReading().`await`()
    XCTAssertEqual(result, NSNumber(value: ECANCELED))
    XCTAssertEqual(reader.finishedReading.result, NSNumber(value: ECANCELED))
    XCTAssertEqual(reader.state, FBFileReaderState.finishedReadingByCancellation)

    pipe.fileHandleForWriting.closeFile()
  }

  func testPipeClosingBehindBackOfConsumer() throws {
    let pipe = Pipe()
    let consumer = FBDataBuffer.accumulatingBuffer()
    let reader = FBFileReader(fileDescriptor: pipe.fileHandleForReading.fileDescriptor, closeOnEndOfFile: false, consumer: consumer, logger: nil)
    XCTAssertEqual(reader.state, FBFileReaderState.notStarted)

    try reader.startReading().`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.reading)

    let expected = "Foo Bar Baz".data(using: .utf8)!
    pipe.fileHandleForWriting.write(expected)
    pipe.fileHandleForWriting.closeFile()
    let predicate = NSPredicate { _, _ in
      expected == consumer.data()
    }
    let expectation = self.expectation(for: predicate, evaluatedWith: self, handler: nil)
    wait(for: [expectation], timeout: FBControlCoreGlobalConfiguration.fastTimeout)

    let result: NSNumber = try reader.stopReading().`await`()
    XCTAssertEqual(result, 0)
    XCTAssertEqual(reader.finishedReading.result, 0)
    XCTAssertEqual(reader.state, FBFileReaderState.finishedReadingNormally)
  }

  func testReadsFromFilePath() throws {
    let reader: FBFileReader = try FBFileReader.reader(withFilePath: "/dev/urandom", consumer: self, logger: nil).`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.notStarted)

    try reader.startReading().`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.reading)

    let result: NSNumber = try reader.stopReading().`await`()
    XCTAssertEqual(result, NSNumber(value: ECANCELED))
    XCTAssertEqual(reader.finishedReading.result, NSNumber(value: ECANCELED))
    XCTAssertEqual(reader.state, FBFileReaderState.finishedReadingByCancellation)
  }

  func testReadingTwiceFails() throws {
    let reader: FBFileReader = try FBFileReader.reader(withFilePath: "/dev/urandom", consumer: self, logger: nil).`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.notStarted)

    try reader.startReading().`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.reading)

    XCTAssertThrowsError(try reader.startReading().`await`())
    XCTAssertEqual(reader.state, FBFileReaderState.reading)

    let result: NSNumber = try reader.stopReading().`await`()
    XCTAssertEqual(result, NSNumber(value: ECANCELED))
    XCTAssertEqual(reader.finishedReading.result, NSNumber(value: ECANCELED))
    XCTAssertEqual(reader.state, FBFileReaderState.finishedReadingByCancellation)
  }

  func testStoppingTwiceDoesNotError() throws {
    let reader: FBFileReader = try FBFileReader.reader(withFilePath: "/dev/urandom", consumer: self, logger: nil).`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.notStarted)

    try reader.startReading().`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.reading)

    var result: NSNumber = try reader.stopReading().`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.finishedReadingByCancellation)
    XCTAssertEqual(result, NSNumber(value: ECANCELED))

    result = try reader.stopReading().`await`()
    XCTAssertEqual(result, NSNumber(value: ECANCELED))
    XCTAssertEqual(reader.finishedReading.result, NSNumber(value: ECANCELED))
    XCTAssertEqual(reader.state, FBFileReaderState.finishedReadingByCancellation)
  }

  func testCancellationOnFinishedReading() throws {
    let reader: FBFileReader = try FBFileReader.reader(withFilePath: "/dev/urandom", consumer: self, logger: nil).`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.notStarted)

    try reader.startReading().`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.reading)

    let finished = reader.finishedReading
    XCTAssertEqual(finished.state, FBFutureState.running)
    try finished.cancel().`await`()
    XCTAssertEqual(finished.state, FBFutureState.cancelled)
    XCTAssertEqual(reader.state, FBFileReaderState.finishedReadingByCancellation)
  }

  func testConcurrentAttachmentIsProhibited() throws {
    let reader: FBFileReader = try FBFileReader.reader(withFilePath: "/dev/urandom", consumer: self, logger: nil).`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.notStarted)

    let concurrentQueue = DispatchQueue.global(qos: .userInitiated)
    let group = DispatchGroup()
    var firstAttempt: FBFuture<NSNull>!
    var secondAttempt: FBFuture<NSNull>!
    var thirdAttempt: FBFuture<NSNull>!

    group.enter()
    concurrentQueue.async {
      firstAttempt = reader.startReading()
      group.leave()
    }
    group.enter()
    concurrentQueue.async {
      secondAttempt = reader.startReading()
      group.leave()
    }
    group.enter()
    concurrentQueue.async {
      thirdAttempt = reader.startReading()
      group.leave()
    }
    group.wait()

    try? firstAttempt.`await`()
    try? secondAttempt.`await`()
    try? thirdAttempt.`await`()
    XCTAssertEqual(reader.state, FBFileReaderState.reading)

    var successes: UInt = 0
    if firstAttempt.state == FBFutureState.done {
      successes += 1
    }
    if secondAttempt.state == FBFutureState.done {
      successes += 1
    }
    if thirdAttempt.state == FBFutureState.done {
      successes += 1
    }

    XCTAssertEqual(successes, 1)
  }

  func testAttemptingToReadAGarbageFileDescriptor() throws {
    let reader = FBFileReader(fileDescriptor: 92123, closeOnEndOfFile: false, consumer: self, logger: nil)
    XCTAssertEqual(reader.state, FBFileReaderState.notStarted)

    try reader.startReading().`await`()

    let result: NSNumber = try reader.stopReading().`await`()
    XCTAssertEqual(result, NSNumber(value: EBADF))
    XCTAssertEqual(reader.finishedReading.result, NSNumber(value: EBADF))
    XCTAssertEqual(reader.state, FBFileReaderState.finishedReadingInError)
  }

  // MARK: FBDataConsumer

  func consumeEndOfFile() {
    didRecieveEOF = true
  }

  func consumeData(_ data: Data) {}
}
